src/internal/common/network/retry_handler.go

```diff
@@ -0,0 +1,186 @@
+package network
```

Comment:
package naming nit: this isn't a strictly network-based paradigm, right?  There's a loose relationship given the return of `headers` in case of range support.  But that doesn't necessarily need to be a header, so much as a k:v tuple.  Additionally, I'm concerned that there's overlap/conflict between this reader handling and the notion of common networking funcs that we may move here in the future.

If we're able to re-use this wrapper in non-network settings, maybe a different package name is in order.  `Readers` or `RetryReader` perhaps.